coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -269,45 +305,149 @@ class SingleClusterPlanner(val dataset: Dataset,
-      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
-      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp)
+      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp, forceInProcess)
+      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp, forceInProcess)
+      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp, forceInProcess)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp, forceInProcess)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+  /**
+   * Returns the shards spanned by a LogicalPlan's data.
+   * @return an occupied Option iff shards could be determined for all leaf-level plans.
+   */
+  private def getShardSpanFromLp(lp: LogicalPlan,
+                                 qContext: QueryContext): Option[Set[Int]] = {
+
+    case class LeafInfo(lp: LogicalPlan,
+                        filters: Set[ColumnFilter],
+                        targetSchemaLabels: Option[Seq[String]])
+
+    // construct a LeafInfo for each leaf...
+    val targetSchemaProvider = qContext.plannerParams.targetSchemaProvider
+    val leafInfos = new ArrayBuffer[LeafInfo]()
+    for (leafPlan <- findLeafLogicalPlans(lp)) {
+      val filters = getColumnFilterGroup(leafPlan)
+      assert(filters.size == 1, s"expected leaf plan to yield single filter group, but got ${filters.size}")
+      leafPlan match {
+        case rs: RawSeries => {
+          // Get time params from the RangeSelector, and use them to identify a TargetSchemaChanges.
+          val tsLabels: Option[Seq[String]] = if (targetSchemaProvider.isDefined) {
+            val tsFunc = targetSchemaProvider.get.targetSchemaFunc(filters.head.toSeq)
+            rs.rangeSelector match {
+              case is: IntervalSelector => findTargetSchema(tsFunc, is.from, is.to).map(_.schema)
+              case _ => None
+            }
+          } else None
+          leafInfos.append(LeafInfo(leafPlan, filters.head, tsLabels))
+        }
+        // Do nothing; not pulling data from any shards.
+        case sc: ScalarPlan => {}
+        // Note!! If an unrecognized plan type is encountered, this just pessimistically returns None.
+        case _ => return None
```

Comment:
This applies to a couple matchers in this code: I could instead (1) add a case for all possible types, or (2) throw an exception when an unrecognized type is encountered. (1) seems like clutter, and (2) seems unsafe.